apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js

```diff
@@ -119,6 +120,16 @@ async function processDeleteFile(req, res) {
   await deleteFile(file.id, res.locals.authn_user.user_id);
 }
 
+function canDeleteAssessmentInstance(resLocals) {
+  // A user can delete an assessment instance from this page if they are
+  // a) a member of course staff with access to this course instance, and
+  // b) own the assessment instance, either directly or via group membership.
+  return (
+    resLocals.authz_data.has_course_instance_permission_view &&
+    resLocals.authz_result.authorized_edit
+  );
```

Comment:
I believe this does not include users with course access but without course instance access. These users would have permission to create an assessment instance, even though they can't see instances from other students. This suggestion will require some prettifying.

```suggestion
  // A user can delete an assessment instance from this page if they are
  // a) a member of course staff with access to this course or course instance, and
  // b) own the assessment instance, either directly or via group membership.
  return (
    (
      resLocals.authz_data.has_course_permission_preview ||
      resLocals.authz_data.has_course_instance_permission_view
    ) &&
    resLocals.authz_result.authorized_edit
  );
```

A consideration in that case, though, is a user that is a student in one term, and then becomes staff in a following term. We might need to ensure that these users cannot mess up with their own assessments when they become staff, though this is already the case right now (as I just raised on Slack).

Reply:
Based on our meeting today: we don't want to give staff without course instance permissions access to deleting their own assessment, as this may give them permission to delete their instance from when they were a student. This PR would still provide a new permission, which is to allow a user that has "course instance student data view" permission to delete their own instance (currently you need student data editor to delete an instance in the students tab). So this suggested change is not necessary (unless you want to add a note that course permission is not enough on purpose).

Reply:
Would it make sense to restrict deletion abilities to only users with `has_course_instance_permission_edit`? This should make this PR not change what's possible at all; it just allows for a shortcut for what can already be done on the Students tab with the same permissions checks.

Reply:
One use-case I can think of is a TA with view permissions that is asked to proofread an assessment. This TA may want to test multiple variants of the same assessment. They may not be able to do this now, but adding that feature may make sense in this context.

Reply:
Mmh, that's a good point. I think as long as we can spell out this justification in a comment, we'll be ok. I'll work on writing this out.

Reply:
Documented in d22666f.